### DISCUSSION TITLE
Script to calculate the best possible layout

### DISCUSSION THREAD

**Discussion by Glitchy-Tozier** [8VIM/8VIM#138] [Show and tell]
(Code: https://github.com/Glitchy-Tozier/8vim_keyboard_layout_calculator)

So I created a small program, which can calculate the best layouts.

### Screenshot of what the results look like:
(Results for a German layout)
![results-shot](https://user-images.githubusercontent.com/59611881/103307486-76eb5b00-4a10-11eb-9d65-6a6ae70e70ae.png)

### To keep this post clean, here's a few links:
- Previous posts about this script (containing some of the thought-process): https://github.com/Glitchy-Tozier/8vim_keyboard_layout_calculator/blob/main/first%20posts%20about%20this%20script.md
- How the program's workflow works: https://github.com/Glitchy-Tozier/8vim_keyboard_layout_calculator/blob/main/Diagrams/How_the_layout_calculator_works.jpg
- You want to influence the future layouts? Please tell me what movements feel comfortable to you here: https://github.com/flide/8VIM/discussions/140

It is pretty much done! :)
The program can find layouts that are (probably) the best possible ones. ((you can never be 100% sure, unless you brute-force everything completely... which is too much work for a computer.

A few small things that still need to be done:
- [x] Make program able create layouts that contain non-ASCII-characters
- [ ] Include the diacritics-gesture in the optimization.
- [x] Clean up the code a little.

For now though, I'll leave it the way it is, until I have tested the layout it currently said was the best (German) one.

**Comment by m-rey** [comment]
Awesome work @Glitchy-Tozier !
I'll take a look at the code, once I have the time

**Reply by Glitchy-Tozier** [reply]
Amazing, thank you!

I've never written a program that needs to be optimized for speed, so please tell me if you see something ridiculously stupid.

**Reply by Glitchy-Tozier** [reply]
Actually, don't look at it yet!
I've changed it a lot and it's no longer what's on Github. Once I've reached the next step in my plan, I'll upload something here.

**Reply by Glitchy-Tozier** [reply]
Now you may look at it  :)

**Comment by sslater11** [comment]
It's a pretty cool script. I'll be honest though, I have no idea how it works haha.

I don't really understand how you're testing the flow, mainly because I don't know what you mean when you say a letter is 'even' or 'odd'
Using the current 8vim layout, I'm guessing that e = 0, i = 1, t = 2, s = 3, etc?

Then I'm assuming that swiping "e t", if e is even and 0, t is 1 step away, and that makes it false.
Then "e y" would be y with 2 steps away, making it false,
Then "e n" would be n with 3 steps away, making it true.
Then "e E" would be E with 4 steps away, making it false? But why? it's a nice loop.
```
#                  -7     -6     -5    -4     -3    -2    -1   ~0~     1      2      3     4      5     6     7
evenNumberFlow = [False, False, True, False, True, True, True, True, False, False, True, False, True, True, True]
```

Something I found whilst making my layout was that there will be times where you may need to sacrifice an inner layer flow to allow more flow in the outer layers, it'll be fun to see how your script handles all of this, or if we just end up with multiple layouts to make an intelligent choice from.

**Comment by Glitchy-Tozier** [comment]
Thanks for checking it out!  :)

>It's a pretty cool script. I'll be honest though, I have no idea how it works haha.

Yeah, there's a lot of stuff in there... It should start making more sense soon. Hopefully....

>I don't really understand how you're testing the flow, mainly because I don't know what you mean when you say a letter is 'even' or 'odd'
Using the current 8vim layout, I'm guessing that e = 0, i = 1, t = 2, s = 3, etc?

What I meant was "even NUMBER" and "odd NUMBER". (No idea whether that's the correct term but even I meant a number that is a multiple of two. (zero counts as even)
And yeah, you got it! `e = 0, i = 1, t = 2, s = 3, etc`!

>Then I'm assuming that swiping "e t", if e is even and 0, t is 1 step away, and that makes it false.
Then "e y" would be y with 2 steps away, making it false,
Then "e n" would be n with 3 steps away, making it true.
Then "e E" would be E with 4 steps away, making it false? But why? it's a nice loop.

I don't quite understand what you mean. After typing zero ('**e**'), the 0th position after that  has a good flow.
```
~0~  
True
```
'**e i**' is horrible. It's **+1** clockwise.
```
  1  
False
```
'**e n**' is **+6** (clockwise) or **-2** (counterclockwise) away.
```
  6
True
```
```
 -2
True
```

Does that make sense?


>Something I found whilst making my layout was that there will be times where you may need to sacrifice an inner layer flow to allow more flow in the outer layers, it'll be fun to see how your script handles all of this, or if we just end up with multiple layouts to make an intelligent choice from.

Yes, this is important.
I'm working on something, I'll post about it when I know it can be done.

ALSO: for me being able to know what I _can_ sacrifice, I need to know how to weigh different factors. This is why I created this question: https://github.com/flide/8VIM/discussions/140

**Comment by sslater11** [comment]
Oh I see, so the numbers relate to the letters going clockwise. I was thinking positive numbers meant "swiping in clockwise direction", and negative numbers meant "swiping in couter-clockwise direction".
Haha it's so easy to misunderstand with this keyboard.

**Reply by Glitchy-Tozier** [reply]
xD Yeah

**Reply by Glitchy-Tozier** [reply]
Btw., something amazing you could do is tell me what movements feel good to you and which you don't like.
(Here: https://github.com/flide/8VIM/discussions/140)

**Comment by Glitchy-Tozier** [comment]
New Update  ^

**Comment by lukamrko** [comment]
Hello,
I'm trying to use your program to generate layout, but I'm constantly getting the following error:

`Traceback (most recent call last):
  File ".\8vim_keyboard_layout_calculator.py", line 1077, in <module>
    main()
  File ".\8vim_keyboard_layout_calculator.py", line 343, in main
    showDataInTerminal(finalLayoutList, finalScoresList, customLayoutNames, customSizeLayouts, customScores, perfectLayoutScore, showData, showGeneralStats, nrOfTopLayouts, nrOfBottomLayouts)
  File ".\8vim_keyboard_layout_calculator.py", line 992, in showDataInTerminal
    print(layoutVisualisation(layout))
UnicodeEncodeError: 'cp932' codec can't encode character '\u27cd' in position 6: illegal multibyte sequence`

The file that I'm trying to use is this one [croatianBigramsLimited.txt](https://github.com/flide/8VIM/files/8163682/croatianBigramsLimited.txt). The way I generated the file is in the following repository: https://github.com/lukamrko/croatianBigramCounter

I'm using pypy3 as instructed. If you need additional info feel free to ask.

**Reply by Glitchy-Tozier** [reply]
Hi, I'm kind of puzzled by that error, but I'll do my best to fix it.

What OS are you using and what language have you set it to?

**Reply by Glitchy-Tozier** [reply]
(@-ing you in case you somehow did't receive the message: @lukamrko)

**Reply by lukamrko** [reply]
Sorry for some reason I didn't get notification in time.
I'm using windows 10 and it is using Japanese language for system locale

**Reply by Glitchy-Tozier** [reply]
@lukamrko Okay, so my first guess is that Windows-console just hates special characters (the more diagonal forward- and back-slashes). Therefore, I tried to replace `⟍ & ⟋` with `\ & /` if the OS is Windows.

Does this version of the script work for you?
[newScript.zip](https://github.com/flide/8VIM/files/8187670/newScripr.zip)

**Reply by lukamrko** [reply]
If I keep using Japanese locale that version doesn't work. But if I switch back to the Croatian locale then that new script works, but the old script still doesn't work. 
New version might not work with Japanese locale because they use yen symbol (¥) instead of  \ or /

EDIT: I managed to make it work in old version too. I just copy-pasted your code with Croatian locale and it generated file nicely.

**Reply by Glitchy-Tozier** [reply]
>  New version might not work with Japanese locale because they use yen symbol (¥) instead of \ or /

What do you mean by that?

> EDIT: I managed to make it work in old version too. I just copy-pasted your code with Croatian locale and it generated file nicely.

So switching to Croation locale allowed you to successfully run the old version?

**Reply by lukamrko** [reply]
> What do you mean by that?

Honestly Japanese version of system locale is too weird and I couldn't get it to work properly. I tried looking at Japanese forums but I couldn't manage to fix it in the end.

>So switching to Croatian locale allowed you to successfully run the old version?

Switching to Croatian locale AND making sure that I put your following code at the end:
 
      if platform.system() is 'Windows':
          blueprint = blueprint.replace('⟍', '\\')
          blueprint = blueprint.replace('⟋', '/')

I made a thread for Croatian layout ([croatian Layout](https://github.com/flide/8VIM/discussions/257)) all thanks to you.

**Reply by Glitchy-Tozier** [reply]
> Switching to Croatian locale AND making sure that I put your following code at the end:
...

I see, so the fix *is* necessary for Windows. Will update the official codebase accordingly!

> I made a thread for Croatian layout (croatian Layout) all thanks to you.

Good job! Mind if I copy your results to this thread? https://github.com/flide/8VIM/discussions/254

**Reply by lukamrko** [reply]
> Good job! Mind if I copy your results to this thread? #254

Yeah feel free. Will add last lines of the evaluation in my original thread soon.

**Comment by Glitchy-Tozier** [comment]
@kjoetom I have further improved a few things.

The percentages are gone, and instead we "only" have access to the scores. This is because scores were completely unrepresentative of how good a layout was. Even languages that didn't have high scores got decent percentages. 

For example, the Icelandic language in an English Layout, had a score of about **50** (compared to optimized scores like **75**) because some of the most important letters could not be typed.
However, its percentage would have been **83.64%**, (optimized = **85%**,) which fooled the user into thinking that the layout was a decent choice for Icelandic.

**Reply by kjoetom** [reply]
Thank you for actively developing the script. I will definitely test it when I am back home after easter and have access to my PC. 

I have also noticed that the scores are not comparable with each other. For example, if you take only 80% of all German trigrams as a basis, you suddenly get scores that are over 100%.

That's why I already tried to make an additional comparison by counting the number of "fullstops" and the number of direction changes (clockwise / counterclockwise).

**Reply by Glitchy-Tozier** [reply]
> I have also noticed that the scores are not comparable with each other. For example, if you take only 80% of all German trigrams as a basis, you suddenly get scores that are over 100%.

1. Are you referring to the old percentages or to the actual scores?
2. I don't quite understand what you did to get >100%
3. This might have been a typo, but you shouldn't use TRIgrams (=three-letter combinations) with the script. Only use BIgrams (=two-letter combinations).
I know there's the option "NGRAM_LENGTH", which is very confusing and unnecessqry, so I'll remove it shortly.

**Comment by kjoetom** [comment]
Hello @Glitchy-Tozier and all of you,

I think I've finally managed to find a score that matches my expectations.

I have explained in another thread (https://github.com/flide/8VIM/discussions/260) how this score is basically built and how to calculate it.

I would like to provide here the score list that can be used in the 8vim_keyboard_layout_calculator. The list consists of 2 parts: 

The first one "SCORES TO BE USED FOR CALCULATING THE BEST LAYOUT" contains ...
... Glitchy's score (ORIGINAL_SCORE_LIST) and 
... my new score (KJOETOM_SCORE_LIST_NEW).

The second part "JUST for COMPARISON OF CUSTOM LAYOUTs" is only for comparing different custom layouts. It is NOT suitable for creating new best layouts, because better layouts get lower scores here. But they are intended to be used to ...
... KJOETOM_DURATION --> calculate the pure duration of writing all bigrams of a set
... KJOETOM_FORCED_STOPS --> count the "forced stops" ("fullstops")
... KJOETOM_DOUBLE_ROT_CHANGES --> count the not so easy to write "RcR-Transitions"
The meaning of these terms is explained in https://github.com/flide/8VIM/discussions/260

I hope that with the tools in the score list attached below, we have a good basis for deciding on the "best layout" for each language.
Here it is: [score_lists.py.tar.gz](https://github.com/flide/8VIM/files/9015481/score_lists.py.tar.gz)

I have also posted a comparison of different German layouts in the thread "German Layout" (https://github.com/flide/8VIM/discussions/90#discussioncomment-3052711).

kjoe

**Reply by Glitchy-Tozier** [reply]
When you're done changing your scores, you could create a pull-request on the optimizer's repo.

It should be doable by just point&clicking: Navigate to the scores-file, hit the edit-button (this will fork the project), then copy your new code into it. Save the file, then github will probably automatically suggest you make a pull-request.

**Reply by kjoetom** [reply]
I will try this out soon. Thanks for the guidance. I am curious whether it is really so easy ;-)

**Reply by kjoetom** [reply]
Just done. This was really easy.

**Reply by Glitchy-Tozier** [reply]
Good job! :)
